keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md

```diff
@@ -0,0 +1,606 @@
+it to determine if this GroupVersionKind is supported.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently the overall design is faulty as `kubeadm init`, despite supporting a version skew with the kubelet
+of N-1, writes a single ConfigMap `kubelet-config-x.y` where `x.y` is the version of the control plane
+and not the version of the kubelet on the host where `kubeadm init` is called.
+
+Later, during `join` operations the version `x.y` is determined by looking up the control plane version
+of the cluster, via the `kubernetesVersion` field of the `ClusterConfiguration` object stored
+in the `kube-system/kubeadm-config` ConfigMap.
+
+During `upgrade` a new `kubelet-config-x.y+1` ConfigMap and RBAC rules are written where `y+1` is the target
+MINOR upgrade version. The old `kubelet-config-x.y` are not cleaned up.
+
+Encoding the version in the API object names is redundant and creates unnecessary complications for both
+kubeadm maintainers and users that wish to modify the object in the cluster.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Simplify the name of the kubelet related ConfigMap and RBAC rules that kubeadm manages to exclude
+encoded version information in the API object names.
+- Remove confusion around the version that is encoded in the name - control plane vs kubelet version?
+- Simplify user access to the ConfigMap, allowing them to always rely on `kube-system/kubelet-config`
+as the default location.
+- Simplify the kubeadm logic to not manage the `x.y` version during PUT/GET operations.
+- Gradually apply the change to allow users to adapt using a feature gate.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Apply the change within a single release of kubeadm, breaking the users.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal is to declare the default ConfigMap and RBAC object format of `kubelet-config-x.y`
+deprecated but support it for a GA deprecation period of 1 year (3 kubeadm releases). In parallel,
+introduce objects named `kubelet-config` (without `x.y`) that will become the source of truth.
```

Comment:
```suggestion
deprecated but support it for a GA deprecation period of 1 year (3 kubeadm releases). In parallel,
introduce an object named `kubelet-config` (without `x.y`) that will become the source of truth.
```

Reply:
there are multiple objects - ConfigMaps and RBAC rules.